src/mgmt/srv_target.c

```diff
@@ -336,6 +339,160 @@ cleanup_leftover_pools(bool zombie_only)
+	bool			 dummy_cancel_state = false;
+	int			 rc;
+	int			 fd;
+	struct stat		 statbuf;
+
+	D_ASSERT(bio_nvme_configured(SMD_DEV_TYPE_META));
+
+	/** generate path to the target directory */
+	rc = ds_mgmt_tgt_file(pool_uuid, NULL, NULL, &pool_path);
+	if (rc) {
+		D_ERROR("newborn path_gen failed for "DF_UUID": "DF_RC"\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		goto out;
+	}
+
+	/** Skip recreation if directory already exists */
+	rc = stat(pool_path, &statbuf);
+	if ((rc == 0) && (statbuf.st_mode & S_IFDIR))
+		goto out;
+
+	/** create the pool directory under NEWBORNS */
+	rc = path_gen(pool_uuid, newborns_path, NULL,
+		      NULL, &pool_newborn_path);
+	if (rc) {
+		D_ERROR(DF_UUID": path_gen failed for: "DF_RC"\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		goto out;
+	}
+	rc = mkdir(pool_newborn_path, 0700);
+	if (rc < 0 && errno != EEXIST) {
+		rc = daos_errno2der(errno);
+		D_ERROR("failed to created pool directory: "DF_RC"\n",
+			DP_RC(rc));
+		/* avoid tgt_destroy(), nothing to do */
+		D_FREE(pool_newborn_path);
+		goto out;
+	}
+
+	/** create VOS files */
+	rc = tgt_vos_preallocate_parallel(pool_uuid, scm_size, tgt_nr,
+					  &dummy_cancel_state);
+	if (rc) {
+		D_ERROR(DF_UUID": failed to create tgt vos files: "DF_RC"\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		goto out;
+	}
+
+	if (rdb_blob_sz) {
+		rc = path_gen(pool_uuid, newborns_path, RDB_FILE"pool",
+			      NULL, &rdb_path);
+		if (rdb_path == NULL) {
+			D_ERROR(DF_UUID": cannot retrieve rdb file info: "DF_RC"\n",
+				DP_UUID(pool_uuid), DP_RC(rc));
+			rc = -DER_NONEXIST;
+			goto out;
+		}
+		fd = open(rdb_path, O_RDWR|O_CREAT, 0600);
+		if (fd < 0) {
+			rc = daos_errno2der(errno);
+			D_ERROR("failed to create/open the vos file %s:"DF_RC"\n",
+				rdb_path, DP_RC(rc));
+			goto out;
+		}
+		rc = fallocate(fd, 0, 0, rdb_blob_sz);
+		close(fd);
+		if (rc) {
+			rc = daos_errno2der(errno);
+			D_ERROR("fallocate on rdb file %s failed:"DF_RC"\n",
+				rdb_path, DP_RC(rc));
+			goto out;
+		}
+		D_FREE(rdb_path);
+	}
+
+	/** move away from NEWBORNS dir */
+	rc = rename(pool_newborn_path, pool_path);
+	if (rc < 0) {
+		rc = daos_errno2der(errno);
+		D_ERROR(DF_UUID": failed to rename pool directory: "DF_RC"\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		goto out;
+	}
+
+	/** make sure the rename is persistent */
+	(void)dir_fsync(pool_path);
+
+out:
+	D_FREE(pool_newborn_path);
+	D_FREE(pool_path);
+
+	return rc;
+}
+
+static int
+recreate_pooltgts()
+{
+	struct smd_pool_info    *pool_info = NULL;
+	struct smd_pool_info    *tmp;
+	d_list_t                 pool_list;
+	int			 rc = 0;
+	int			 pool_list_cnt;
+	daos_size_t		 rdb_blob_sz = 0;
+	struct d_uuid		 id;
+
+	D_ASSERT(bio_nvme_configured(SMD_DEV_TYPE_META));
+	D_INIT_LIST_HEAD(&pool_list);
+	rc = smd_pool_list(&pool_list, &pool_list_cnt);
+	if (rc != 0) {
+		D_ERROR("Failed to get pool info list from SMD\n");
+		return rc;
+	}
+
+	d_list_for_each_entry_safe(pool_info, tmp, &pool_list, spi_link) {
+		/* Cleanup Newborns */
```

Comment:
Usually the blobs for pool in ZOMBIES should have been destroyed, but if the pool is still referenced on destroy, the blobs destroy will be deferred to ZOMBIES cleanup on next restart (or next pool creation). See tgt_destroy() -> tgt_kill_pool() -> vos_pool_kill(). But that was implemented long time ago, I think today we've ensured the pool is not referenced on destroy?

Anyway, let's keep it in mind, we could address it later if we really hit the issue.